README.md
# Leaflet.GeojsonLayerSwitcher [![NPM version](https://badge.fury.io/js/leaflet-geojson-layer-switcher.svg)](https://npmjs.org/package/leaflet-geojson-layer-switcher) [![Build Status](https://api.travis-ci.org/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher.svg?branch=master)](https://travis-ci.org/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher)

This plugin is intended to be used for Polyline selection. Navigates between layers and switches the selection state.

## Installation

### npm

```sh
$ npm install --save leaflet-geojson-layer-switcher
```

### browser

- Download or clone this github project
- Include `dist/Leaflet.GeojsonLayerSwitcher.min.js` and `dist/Leaflet.GeojsonLayerSwitcher.min.css` in your project

### compile

- Download or clone this github project
- Run `npm i` in the project folder
- Run `npm run build`

## Demo

![demo picture](./assets/demo.png)

## Usage

### es6 include

```js
import L from 'leaflet';
import leafletGeojsonLayerSwitcher from 'leaflet-geojson-layer-switcher';
leafletGeojsonLayerSwitcher(L);
```

Don't forget to link the CSS file.

### non-es6 include

```html
<link rel="stylesheet" href="Leaflet.GeojsonLayerSwitcher.min.css" />
<script src="Leaflet.GeojsonLayerSwitcher.min.js"></script>
```

```js
require('js/Leaflet.GeojsonLayerSwitcher.js').default(L);
```

### Interaction

```js
const map = L.map('map', {
    center: [45.1834782, 5.7831946],
    zoom: 13
});

L.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png').addTo(map);

const geojson = {
	type: "FeatureCollection",
	features: [
		{
			type: "Feature",
			properties: {
				enabled: true
			},
			geometry: {
				type: "LineString",
				coordinates: [[5.0,45.0],[5.1,45.1],[5.0,45.2]]
			}
		},
		{
			type: "Feature",
			properties: {
				enabled: false
			},
			geometry: {
				type: "LineString",
				coordinates: [[6.0,46.0],[6.1,46.1],[6.0,46.2]]
			}
		},
		{
			type: "Feature",
			properties: {},
			geometry: {
				type: "LineString",
				coordinates: [[7.0,47.0],[7.1,47.1],[7.0,47.2]]
			}
		}
	]
};

const handler = L.geoJSONLayerSwitcher(map, {
	propertyStateSelected: 'enabled'
});

handler.enable();
handler.addData(geojson);

// USER INTERACTS WITH PLUGIN

const selectedLayersArray = handler.getSelection();
handler.disable();

//Do whatever you want with your layers.
```

## API

### Factory `L.geoJSONLayerSwitcher(<Map> map, <Object> options?)`

#### Options
- propertyStateSelected `<String>` : The feature's property name containing a boolean defining wether the layer is selected. If feature does not have this property, it will be initialized to `false`. Default property is `'selected'`.
- style `<Object>` : Customize style applied to polylines which are not selected. Default is dashed and red.
- styleSelected `<Object>` : Customize style applied to selected polylines. Default is dashed and green.
- completeButton `<Boolean>` : Add a complete button in the popup. When clicked, it will trigger a `layerSwitcher.complete` map event container the selected layers
- popupAnchor `<Array>` : Allows you to change Leaflet.Dialog anchor. Defaults to `[10, 60]`.
- position `<String>` : Allows you to change Leaflet.Dialog position. Defaults to `topright`.

Note : style and styleSelected support PolylineDecorator. Just provide the `patterns` key any configuration needed from PolylineDecorator.

## License

MIT © [Easy-Mountain](https://github.com/easymountain)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.2](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.1.1...v1.1.2) (2023-01-30)

### [1.1.1](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.1.0...v1.1.1) (2023-01-17)

## [1.1.0](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.0.10...v1.1.0) (2023-01-17)


### Features

* add an optional complete button and event ([2e0fa85](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/commit/2e0fa852422099b6dff8b401f435b4afe70e6c3c))

### [1.0.10](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.0.8...v1.0.10) (2023-01-17)

### [1.0.9](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.0.8...v1.0.9) (2023-01-17)

### [1.0.8](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/compare/v1.0.7...v1.0.8) (2022-10-17)


### Bug Fixes

* add position to dialog ([1277533](https://github.com/EASYMOUNTAIN/Leaflet.GeojsonLayerSwitcher/commit/1277533ecadeef77891733639e6ea29d217558dc))

### 1.0.7 (2022-10-17)

## [1.0.0] - 2018-12-04
- Plugin drop

[1.0.0]: https://github.com/easymountain/leaflet-geojson-layer-switcher/compare/1.0.0...master


src/js/Leaflet.GeojsonLayerSwitcher.js
export default function(L) {

	const commonStyle = {
		opacity: 1,
		weight: 5,
		dashArray: '10, 30',
	};

	const GeoJSONLayerSwitcher = L.GeoJSONLayerSwitcher = L.Handler.extend({

		options: {
			propertyStateSelected: 'selected',
			style: Object.assign({color: '#f00'}, commonStyle),
			styleSelected: Object.assign({color: '#0f0'}, commonStyle),
			popupAnchor: [10, 60],
			position: 'topright',
			completeButton: false,
		},

		initialize: function(map, options) {
			L.Handler.prototype.initialize.call(this, map);
			L.Util.setOptions(this, options || {});

			const context = this;

			this._dialog = L.control.dialog({
				initOpen: false,
				size: [300, 160],
				anchor: context.options.popupAnchor,
				position: context.options.position,
			});

			this._geoJSONLayer = L.geoJson(null, {
				style: function(feature) {
					if (!feature.properties.hasOwnProperty(context.options.propertyStateSelected)) {
						feature.properties[context.options.propertyStateSelected] = false;
					}
					return context._getStyleFor(feature);
				},
				onEachFeature: function(feature, layer) {
					const style = context._getStyleFor(feature);
					if (L.PolylineDecorator && style.hasOwnProperty('patterns')) {
						layer.decorator = L.polylineDecorator(layer, {
							patterns: style.patterns
						});
						layer.decorator.addTo(context._map);
					}
				}
			});
		},

		_getStyleFor: function(feature) {
			return (feature.properties && feature.properties[this.options.propertyStateSelected]) ? this.options.styleSelected : this.options.style;
		},
		addHooks: function() {
			L.DomEvent.on(document, 'eventname', this._doSomething, this);
		},

		removeHooks: function() {
			L.DomEvent.off(document, 'eventname', this._doSomething, this);
		},
		addHooks: function() {
			this._geoJSONLayer.addTo(this._map);
			this._dialog.addTo(this._map);
			const dialogContent = this._initDialogContent();
			this._dialog.setContent(dialogContent);
		},

		removeHooks: function() {
			this._geoJSONLayer.getLayers().forEach(layer => {
				if (L.PolylineDecorator && layer.hasOwnProperty('decorator')) {
					this._map.removeLayer(layer.decorator);
				}
			});
			this._geoJSONLayer.clearLayers();
			this._map.removeLayer(this._geoJSONLayer);
			this._dialog.close();
			this._map.removeControl(this._dialog);
		},

		addData: function(geojson) {
			this._geoJSONLayer.addData(geojson);
			const layers = this._geoJSONLayer.getLayers();

			this._index = 0;
			this._count = layers.length;
			if (layers.length > 0) {
				this._goTo(layers[0]);
			}
			this._dialog.open();
			return this;
		},

		_initDialogContent: function() {
			let container = L.DomUtil.create('div');

			this._navTitle = L.DomUtil.create('h4', 'leaflet-geoJSONLayerSwitcher-title', container);
			let navBar = L.DomUtil.create('div', 'leaflet-geoJSONLayerSwitcher-navbar', container);

			this._prevButton = L.DomUtil.create('a', 'leaflet-geoJSONLayerSwitcher prev', navBar);
			L.DomUtil.create('i', 'fa fa-chevron-left', this._prevButton);

			this._switchLabel = L.DomUtil.create('label', 'switch', navBar);
			this._switchButton = L.DomUtil.create('input', '', this._switchLabel);
			this._switchButton.setAttribute('type', 'checkbox');

			L.DomUtil.create('span', 'slider round', this._switchLabel);

			this._nextButton = L.DomUtil.create('a', 'leaflet-geoJSONLayerSwitcher next', navBar);
			L.DomUtil.create('i', 'fa fa-chevron-right', this._nextButton);

			const context = this;

			L.DomEvent.on(this._prevButton, 'click', function() {
				context._previous();
			}, this);

			L.DomEvent.on(this._nextButton, 'click', function() {
				context._next();
			}, this);

			if (this.options.completeButton) {
				this._completeButton = L.DomUtil.create('a', 'leaflet-geoJSONLayerSwitcher complete', navBar);
				L.DomUtil.create('i', 'fa fa-check', this._completeButton);

				L.DomEvent.on(this._completeButton, 'click', function(event) {
					context._map.fire('layerSwitcher.complete', {
						originalEvent: event,
						layers: context.getSelection(),
					});
				}, this);
			}

			L.DomEvent.on(this._switchButton, 'click', function() {
				context._toggle();
			}, this);

			return container;
		},

		_index: 0,
		_count: 0,

		_previous: function() {
			this._index = (this._index + this._count - 1) % this._count;
			this._goTo(this._currentLayer());
		},

		_next: function() {
			this._index = (this._index + 1) % this._count;
			this._goTo(this._currentLayer());
		},

		_goTo: function(layer) {
			if (layer === null) {
				return;
			}
			this._focus(layer);
			this._updateDialogFor(layer);
		},

		_toggle: function() {
			let layer = this._currentLayer();
			if (layer === null) {
				return;
			}

			if (L.PolylineDecorator && layer.hasOwnProperty('decorator')) {
				this._map.removeLayer(layer.decorator);
				delete layer.decorator;
			}

			layer.feature.properties[this.options.propertyStateSelected] = !layer.feature.properties[this.options.propertyStateSelected];

			const style = this._getStyleFor(layer.feature);
			layer.setStyle(style);

			if (L.PolylineDecorator && style.hasOwnProperty('patterns')) {
				layer.decorator = L.polylineDecorator(layer, {
					patterns: style.patterns
				});
				layer.decorator.addTo(this._map);
			}

			this._updateToggleButtonFor(layer);
		},

		_updateDialogFor: function(layer) {
			if (layer === null) {
				this._dialog.close();
				return;
			}

			const readableIndex = this._index + 1;
			this._navTitle.innerHTML = readableIndex+'/'+this._count;
			this._updateToggleButtonFor(layer);
		},

		_updateToggleButtonFor: function(layer) {
			this._switchButton.checked = layer.feature.properties[this.options.propertyStateSelected] === false;
		},

		_focus: function(layer) {
			if (layer === null) {
				return;
			}
			this._map.fitBounds(layer.getBounds());
		},

		_currentLayer: function() {
			if (this._count === 0) {
				return null;
			}
			return this._geoJSONLayer.getLayers()[this._index];
		},

		getSelection: function() {
			let selection = [];
			this._geoJSONLayer.getLayers().forEach(layer => {
				if (layer.feature.properties[this.options.propertyStateSelected]) {
					selection.push(layer);
				}
			});
			return selection;
		},
	});

	// Factory
	L.geoJSONLayerSwitcher = function(map, options) {
		return new GeoJSONLayerSwitcher(map, options);
	};

	return GeoJSONLayerSwitcher;
}
